Update module debugging docs (#70847)
ansible/ansible: docs/docsite/rst/dev_guide/debugging.rst

    You can modify files in this directory if you suspect that the module is having a problem in some of this boilerplate code rather than in the module code you have written.

#. Once you edit the code or arguments in the exploded tree, use the ``execute`` subcommand to run it:

    .. code-block:: shell-session

      $ python AnsiballZ_ping.py execute
      {"invocation": {"module_args": {"data": "debugging_session"}}, "changed": false, "ping": "debugging_session"}

    This subcommand takes care of setting the ``PYTHONPATH`` to use the exploded ``debug_dir/ansible/module_utils`` directory and invoking the script using the arguments in the ``args`` file. You can continue to run the module like this until you understand the problem. Then you can copy the changes back into your real module file and test that the real module works via ``ansible`` or ``ansible-playbook``.

> 

Reviewer: @samdoran Can we add some details about custom PYTHONPATH since collections might be installed at a different locations? Also, adding debbuging with Pycharm and VS code will be good to have but not a blocker.

Reviewer 2: > 
> @samdoran Can we add some details about custom PYTHONPATH since collections might be installed at a different locations? Also, adding debbuging with Pycharm and VS code will be good to have but not a blocker.

POV from documenter
Agree for the PYTHONPATH as it makes sense.
About adding information about different editors I'm a bit reluctant, I fear we will have to maintain that for all editors people want to use, and they may be out of date pretty quickly.  
However that's kind of information that may be interesting in a blog post on the Ansible blog for a particular editor.

Reviewer: Make sense.

Reviewer: Scratch the idea of Editor specific debugging steps. @bmillemathias Thanks.

Reviewer 2: > Scratch the idea of Editor specific debugging steps. @bmillemathias Thanks.

And I'm sure the bullhorn team would be more than happy to talk about this future blog post :)

Author: @Akasurde I could mention that other debuggers should work with these methods as well. I assumed that was somewhat implied, but that's a bad assumption on my part. @nitzmahone did a Fest talk a while back on using the PyCharm remote debugger, so it is possible.

Debugging modules
*****************

> .. _debugging_modules:

Reviewer 3: This main page anchor should go above the main page title.

> Ansible modules are put together as a zip file consisting of the module file and the various Python module boilerplate inside of a wrapper script. The file needs to be extracted from the wrapper in order to see what's actually going on in the module. The wrapper script provides some helper methods to do just that.

Reviewer 3: ```suggestion
Ansible modules are put together as a zip file consisting of the module file and the various Python module boilerplate inside of a wrapper script. To see what is actually happening in the module, you need to extract the file from the wrapper. The wrapper script provides helper methods that let you do that.
```

> The follow steps use ``localhost`` as the target host, but they can be used when debugging against remote hosts as well. For a simpler approaches to debugging without using the temporary files, see :ref:`simple debugging <simple_debugging>`.

Reviewer 3: ```suggestion
The following steps use ``localhost`` as the target host, but you can use the same steps to debug against remote hosts as well. For a simpler approach to debugging without using the temporary files, see :ref:`simple debugging <simple_debugging>`.
```

    You can modify files in this directory if you suspect that the module is having a problem in some of this boilerplate code rather than in the module code you have written.

#. Once you edit the code or arguments in the exploded tree, use the ``execute`` subcommand to run it:

    .. code-block:: shell-session

      $ python AnsiballZ_ping.py execute
      {"invocation": {"module_args": {"data": "debugging_session"}}, "changed": false, "ping": "debugging_session"}

    This subcommand takes care of setting the ``PYTHONPATH`` to use the exploded ``debug_dir/ansible/module_utils`` directory and invoking the script using the arguments in the ``args`` file. You can continue to run the module like this until you understand the problem. Then you can copy the changes back into your real module file and test that the real module works via ``ansible`` or ``ansible-playbook``.

.. _simple_debugging:

> Simple Debugging

Reviewer 3: Our style guide calls for sentence case for headings. We're trying (slowly) to update them all.
```suggestion
Simple debugging
```

================

> The easiest way to run a debugger in a module, either local or remote, is to use `epdb <https://pypi.org/project/epdb/>`_. Add ``import epdb; epdb.serve()`` in the module code on the control node at the desired break point. To connect to the debugger, run ``epdb.connect()``. See the documentation for how to specify the ``host`` and ``port``. If connecting to a remote node, make sure to use a port that is allowed by any firewall between the control node and the remote node.

Reviewer 3: Not sure if "the documentation" here means the epdb docs, our docs, or some other resource. I'd add a link, or specify, or both.
